### PR TITLE
feat(aquavm-air,air-interpreter-signature,air-interpreter-data)!: Peer signatures

### PR DESCRIPTION
**Pull request by monoid**
A peer signs call results and canon results it has produced.

New field `signatures`, a map from peer public key to signature, is added to the interpreter data.

Signatures checking is yet to be done.

### CODE REVIEW

File: air/src/execution_step/execution_context/context.rs [hunk 1]
```diff
@@ -69,6 +73,12 @@ pub(crate) struct ExecutionCtx<'i> {
 
     /// CID-to-something trackers.
     pub(crate) cid_state: ExecutionCidState,
+
+    /// Signatures' store.
```
Comment (mikevoronov) [inline air/src/execution_step/execution_context/context.rs:77]: let's add something about why is this field needed

File: air/src/execution_step/execution_context/context.rs [hunk 2]
```diff
@@ -69,6 +73,12 @@ pub(crate) struct ExecutionCtx<'i> {
...
+    pub(crate) signature_store: SignatureStore,
+
+    /// Local signature tracker.
```
Comment (mikevoronov) [inline air/src/execution_step/execution_context/context.rs:80]: ditto

File: crates/air-lib/interpreter-signatures/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+use std::borrow::Borrow;
+use std::collections::HashMap;
+use std::hash::Hash;
+
+/// An opaque serializable representation of public key.
+///
+/// It can be string or binary, you shouldn't care about it unless you change serialization format.
+#[derive(Debug, Hash, Clone, Eq, PartialEq, PartialOrd, Deserialize, Serialize)]
+#[serde(transparent)]
+pub struct PublicKey(Box<str>);
```
Comment (mikevoronov) [inline crates/air-lib/interpreter-signatures/src/lib.rs:44]: let use `Vec<u8>` here

File: crates/air-lib/interpreter-signatures/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+
+impl From<fluence_keypair::PublicKey> for PublicKey {
+    fn from(value: fluence_keypair::PublicKey) -> Self {
+        Self(bs58::encode(&value.to_vec()).into_string().into())
+    }
+}
+
+#[derive(Debug, Hash, Clone, Eq, PartialEq, PartialOrd, Deserialize, Serialize)]
+#[serde(transparent)]
+pub struct Signature(Box<str>);
```
Comment (mikevoronov) [inline crates/air-lib/interpreter-signatures/src/lib.rs:54]: ditto

**Review by mikevoronov — APPROVED** [review]
utACK

File: crates/air-lib/interpreter-signatures/src/lib.rs (hunk 3 above)
Comment (monoid) [reply crates/air-lib/interpreter-signatures/src/lib.rs:44]: They are used as keys in JSON hash maps, and only strings can be used as keys.
This structure is designed to be an opaque type with implemtation can be changed to `Vec<u8>` (or `Box<[u8]>`) when we switch to a binary format.

File: crates/air-lib/interpreter-signatures/src/lib.rs (hunk 4 above)
Comment (monoid) [reply crates/air-lib/interpreter-signatures/src/lib.rs:54]: As we use JSON yet, a string is more effective.  The type is designed to be opaque, and can be easily changed to `Vec<u8>` (or `Box<[u8]>`) when we switch to binary format.

File: crates/air-lib/interpreter-signatures/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+    unused_variables,
+    unused_unsafe,
+    unreachable_patterns
+)]
+
+use air_interpreter_cid::CID;
+use fluence_keypair::error::SigningError;
+use fluence_keypair::KeyPair;
+use rand_chacha::rand_core::SeedableRng;
+use serde::{Deserialize, Serialize};
```
Comment (raftedproc) [inline crates/air-lib/interpreter-signatures/src/lib.rs:33]: IIRC there should be one import per line.

File: crates/air-lib/interpreter-signatures/src/lib.rs (hunk 3 above)
Comment (mikevoronov) [reply crates/air-lib/interpreter-signatures/src/lib.rs:44]: I see, so let's add a comment about that.

**Pull request merged**